### ISSUE TITLE
Asynchronous/tokio ResolvesServerCert?

### ISSUE THREAD

**Issue by alex** [rustls/rustls#89]
I don't have a use case for this yet, but as more and more of the rust community moves towards async libraries like tokio, I wonder if there's a way `ResolvesServerCert` could support callback functions which do not return immediately/play nicely with the `tokio` `Core`.

Specific use case that might require this are servers which process many different domains and store certs/keys in a database, or servers which attempt to lazily obtain certificates from CAs like let's encrypt.

**Comment by ctz** [comment]
Do you have any ideas how the API should look? I haven't looked much at the internals of tokio/futures so it's not totally obvious how to achieve this without having tokio/futures-level concepts spilling "up" into the top-level API.

**Comment by alex** [comment]
I think the two choices are basically:

- have it return a `Future`
- have it take a callback `Fn` as an argument, and it should call that once the result is ready

**Comment by jeromegn** [comment]
We also need this. We're currently using openssl's SNI callback feature (which is synchronous, but if rustls allowed this to be asynchronous, it would be a great argument to switch.) We fetch certificates from a database, and then decrypt them.

Providing a callback `Fn` as an argument is probably the most flexible, but since futures have just been stabilized, it might be more idiomatic to use them. It could be optional as a feature of the crate to retain backward-compatibility.

**Comment by jeromegn** [comment]
I'd very much like rustls to be more asynchronous now that it's a stable rust feature.

I think a simple `ResolvesServerCertAsync` trait with a `resolve` function signature returning a `Pin<Box<dyn Future...` would work fine. It should be implemented for `ResolvesServerCert` by returning a future that returns immediately with the result.

Looking at the code, it appears there are a lot of pieces that would need to be modified to either accept or return futures.

This would also make rustls async by default, a big breaking change. Admittedly, most operations rustls need to perform are synchronous in nature, so maybe this isn't the best way forward.

**Comment by glebpom** [comment]
We also need it. We want to retrieve certificates from another service based on SNI. I'd like to help with that, but it sounds like it requires some high-level decisions, on how to implement it.

**Comment by glebpom** [comment]
From what I see in the code, it looks like making it async is premature. At least async in Traits should be stabilized to prevent performance degradation due to additional allocation. I think the best solution would be to change the public interface of `rustls::ServerSession::new`, so that it returns `ClientHello` under some new struct. The user's code should choose the certificate based on SNI, and invoke later processing. It will make `ResolvesServerCert` trait unnecessary and will be compatible with async certificate retrieval and crates like tokio-rustls.

**Comment by Ralith** [comment]
I think that's an *excellent* idea! It makes the interface both simpler and more flexible.

**Comment by gorup** [comment]
Ready for a "good" hack to get around this quickly? 

0. Read some amount of tcp bytes into a buf, maybe using `.peek()`
   ```
    // let mut connection: TcpStream = ...
    let mut buf = vec![0; 1024];
    let n = connection.peek(&mut buf).await?;
    ```
1. Create a ServerConfig that will accept any connection to anything, any SNI
2. Create a new ServerSession from this config, called `temp_sess`
3. `temp_sess.read_tls(&mut buf[0..n])`
4. Bam, at this point, you can call `get_sni_hostname` and do whatever you want
5. Now if you look up a Cert for this SNI (using your own logic), then you create a new ServerConfig with that good Cert
6. Create the final session from this good ServerConfig, and just read the bytes as you were, e.g. `good_sess.read_tls(&mut buf[..])`
7. Then you can just continue to do things as normal, and at this point you were able to:
  - Read the SNI
   - Decide which ServerConfig to use, asynchronously
   - Continue authorization based on the selected ServerConfig just like normal

It's hacky, and yes you technically are reading the same `&mut buf[..]` bytes 2x and processing them 2x, but this is just the ClientHello so its what, 300 bytes? max?

Anyways, this may help you or others who want to peek the SNI, then do some stuff, then decide to use a specific ServerConfig.

**Comment by jeromegn** [comment]
@gorup we're pretty much doing exactly that :/ it's very hacky but it does work. We're using the `tls-parser` crate though. A `ServerSession` like you're doing might actually be better!

**Comment by gorup** [comment]
> Edit: ClientHello is more than 300 bytes. I've had to read up to 2000 bytes sometimes.

Good to know! I'll need to change some buffer sizes and expectations on my initial reads. Thx @jeromegn

**Comment by djc** [comment]
I do think we need some improvements along these lines (and I'd be happy to work on something). I'm not sure I fully grok @glebpom's proposal around changing `ServerSession::new()` -- I think it assumes you'll always have a full `ClientHello` available, but that's unlikely to be universally true especially as post-quantum crypto comes into play.

Given the use cases from @jeromegn and @glebpom, it sounds like async would be attractive, but using async trait methods, while probably the best solution in the hypothetical future where we have GAT, would incur a performance penalty. I'm wondering if there are alternative solutions here where we might for example parametrize `ServerConfig` with a concrete type that implements `Future`, providing a default that reuses the current infrastructure.

Even then, though, waiting for the full `ClientHello` to be available before kicking off such a future seems like it might be undesirable, but in that case different implementations might want to wait for different parts of data (for example, just SNI, just ALPN, ALPN + SNI) to be available before kicking off the future. So in that case, maybe the trait should tell the `ServerSession` what it needs and then the `ServerSession` will kick off the future as soon as it has parsed all the required pieces out of the `ClientHello`?

**Comment by glebpom** [comment]
I think using some sort of Future in non-future-aware code is problematic because we need to wait for the result in the rustls itself. We will have only one option - to block until the result is ready, which will break the whole idea.
I'm wondering, what is the problem with waiting for full `ClientHello`? I thought that we are unable to proceed to the actual TLS data transmission until we have ClientHello fully received. If it's only about proceeding with the received data as fast as possible, I'm wondering if it's a real problem. It looks like getting a few additional bytes may not introduce such a severe penalty.
Anyway, if it is, we can try to implement the Iterator with reference to partially filled `ClientHello` along with `ServerSession` or some other structure each time when new data arrives. When it's done, the ready session may be retrieved. The code may look like this:
```
let new_session = ServerSession::new(&config);
for (updated_state, session) in new_session.updates() {
   if let Some(hostname) = updated_state.get_sni_hostname() {
      let (key, cert) = certs.get(hostname).unwrap();
      session.set_single_cert(key ,cert).unwrap();
   }
}
let session = new_session.finish();
```
The additional benefit of this approach, is that if we don't need any of extra data, we can just ignore the iterator, and immediately call `finish`.

**Comment by Ralith** [comment]
> Even then, though, waiting for the full ClientHello to be available before kicking off such a future seems like it might be undesirable

I'm not sure it's possible/desirable/useful to yield information prior to complete ClientHello being received. I also strongly feel that explicit async is excess complexity for this case, vs. just exposing enough of a state machine to address this use case in external code whether it's async or not.

I'm imagining an interface that looks something like:

```rust
let acceptor = Acceptor::new(&config);
while acceptor.client_hello().is_none() {
    acceptor.read(&incoming_data);
}
app_specific_handling(acceptor.client_hello().unwrap());
let session = ServerSession::new(acceptor);
```

Note that a handy [`ClientHello`](https://docs.rs/rustls/0.17.0/rustls/struct.ClientHello.html) wrapper is already exposed.

QUIC would need a slightly distinct interface since there's one less layer of framing there, but the principle would be the same.

**Comment by glebpom** [comment]
@Ralith Looks good to me. Do I get it right, that`ServerSession::new(acceptor)` will do all necessary reads until  `ClientHello` is ready even if the user's code avoids waiting for it?

**Comment by djc** [comment]
Your interface proposal looks good to me. @ctz what do you think? I'm happy to implement this.

**Comment by glebpom** [comment]
For those who may be interested in supporting SNI in warp, with the approach mentioned by @gorup, take a look here: https://github.com/glebpom/warp/commit/030e2a89a24e2c8c63a112c210688c76d8e5c42b

**Comment by jeromegn** [comment]
@glebpom Nice work. That's exactly how we do it at Fly.io (using tls-parser too). Has been working well for months, doing thousands of handshakes per second.

According to @gorup you could do it without tls-parser if you don't want to add that dependency.

**Comment by Rantanen** [comment]
@glebpom @jeromegn How reliable the `poll_peek` is? The documentation on `peek` mentions it's implemented by passing `MSG_PEEK` to the underlying `recv` system call. Is the system buffer unbounded/large enough to guarantee the code won't end up in a state where it repeatedly calls `peek` waiting for `ClientHello` to complete but the `recv` buffer is full, never being able to complete the `ClientHello`?

I ended up resolving the same issue by reading the bytes out of the incoming stream and in addition to parsing them (using `ServerSession` in my case), I'm duplicating them into a `Vec<u8>` structure. ([tls::resolve_client_hello](https://github.com/Rantanen/proxide/blob/d04bcda4fc8c57ae0d2a345d13362f2e5b7c421c/src/connection/tls.rs#L182-L192))

After parsing the `ClientHello`, I'm then restarting the `ServerSession` with a [`PrefixdStream`](https://github.com/Rantanen/proxide/blob/d04bcda4fc8c57ae0d2a345d13362f2e5b7c421c/src/connection/stream.rs#L7-L17) that is able to put yield the already-read data in front of the stream.

I'd love to replace this all with `peek` if I had faith badly behaving TLS clients didn't cause issues for it.

**Comment by glebpom** [comment]
@Rantanen I'm growing the buffer in my implementation up to a certain limit until `ClientHello` is successfully parsed. `peek` will always read from the beginning.

**Comment by Rantanen** [comment]
@glebpom You are growing the user mode buffer, but surely that's not the only buffer that holds the data? There must be some sort of kernel buffer as well, which I'd expect to have limits in place. As far as I know, given the presence of extensions (and your need to parse SNI from said extensions), there could be arbitrary amount of data before SNI appears in the `ClientHello`.

I'm sure most "real" TLS clients will send sensibly sized `ClientHello`s, but that implementation looks like it's susceptible to a DoS attack from a client that sends arbitrarily large `ClientHello` that overflows the kernel buffer.

Just confirmed over at playground, the buffer seems to be somewhere around 60 and 100 KB:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4b5cabca3e585939a9f653d65fbd54b3

If the `ClientHello` is larger than that (or more accurately if the SNI is beyond the first 100 KB of data), the implementation in warp will end up timing out on the `poll_peek` (assuming there is a timeout somewhere up in the chain.)

https://github.com/glebpom/warp/blob/030e2a89a24e2c8c63a112c210688c76d8e5c42b/src/tls.rs#L270-L286

It might make sense to cap the size of `buf` and just error out of the loop if it seems like `pin.buf` is about to grow beyond a limit. A sensible client hello should _not_ be multiple kilobytes of data.

(I think the 62 KB lower limit on the `recv` buffer should be enough for parsing even a worst case `ClientHello`, excluding malicious ones, giving me enough confidence to move Proxide on top of a similar implementation.)

**Comment by jameysharp** [comment]
Similarly to @Rantanen in https://github.com/ctz/rustls/issues/89#issuecomment-636188966, I wanted to use the client's SNI to decide which ALPN protocols to accept. I started hacking together a callback-based trait like `ResolvesServerCert` and then realized that I wanted async support, so I started thinking about a typestate-style refactoring of the `ServerSession` API…

What I thought of ended up being almost exactly what @glebpom and @Ralith have described here, so I'd like to give a big :+1: to their proposals.

The reason I think that tracking protocol state at the type level is a good idea is that from the server's point of view there are really three major phases in a TLS connection, I think: Waiting for `ClientHello`, then completing the handshake, then exchanging application data.

These phases are fundamentally different in terms of both state needed and API operations supported, which are both good arguments for representing them with distinct types.

In terms of state, for example, I don't think rustls should keep the `ClientHello` after moving to the handshake phase. In my opinion, the application ought to be responsible for saving whatever information it needs, including SNI and selected ALPN protocol, so it doesn't pay the cost for saving information it doesn't care about. Similarly, most of the `ServerConfig` data is irrelevant once the handshake finishes.

In terms of API operations, there are `ServerSession` methods that only work during the handshake, others that only work after the handshake, and some that return meaningless results until some poorly-defined moment during the handshake. Currently you have to read the documentation to get those right. Further, there are important differences around application data. No application data can be exchanged until after the `ClientHello` has been read. During the final phase, most of the useful API is in exchanging application data via the `Read` and `Write` traits. But in between those two phases, the server can choose to read 0-RTT early data and/or send encrypted data to the not-yet-authenticated peer, but developers should understand the security implications, so making the handshake phase's type directly implement `Read`/`Write` for application-layer data seems unwise.

I think these are exactly the right phase boundaries for all the application-specific decisions a server might want to make, exposing more or less the full flexibility of the TLS spec. Going from handshake to app data just requires approving the client-provided certificate, if there is one, I think. And once you have the `ClientHello`, you can construct a builder-pattern `ServerConfig` that doesn't need any callbacks or trait objects, and use that to start the handshake phase. Since the handshake phase is short, you shouldn't even need reference counting for objects like certificates; just give the type representing this phase a lifetime parameter and let it borrow all the necessary data.

As another bonus, I think these changes should help eliminate a lot of dynamic memory allocations inside rustls, which is usually good for performance as well as being a step toward `no_std` support as in #157 and #283.

There's probably an argument to be made for taking the same approach on the client side, except I think there's just two phases, separating the handshake from application data, optionally choosing a client certificate in between.

**Comment by Matthias247** [comment]
I think @glebpom is fully right in that `Future`s  aren't a good solution here, because they are "infectious" (would require other rustls APIs also being updated), and just extending the state machine and letting wrappers (like tokio-tls and quinn) deal with `Future`s is preferable.

The API in https://github.com/ctz/rustls/issues/89#issuecomment-630776232 might work, but I think this can even be more simplified. `Session` could be extended by a `fn wants_cert()` call, which needs to be checked during handshaking in addition to the existing `wants_read()` and `wants_write()` calls. While the session is stuck on needing a cert the handshake won't make progress (and `wants_read/write` might return `false`). 

Users can then
- Read the parsed `ClientHello` with another method on the `Session`, to determine which cert might be needed
- Feed the cert via an explicit `set_single_cert` call

When this is done `wants_cert()` will return `false` and the handshake can be progress via a call to `process_new_packets` (even though that is a bit misnamed for this purpose).

Ideally the function to set a certificate would support it in a form that allows applications to cache them for the same SNI (which e.g. means an `Arc<Cert>`, and requiring the least amount of work per handshake.

Alternatively there could also be an option to set a callback interface which is invoked when the session retrieves the client hello. That is something that s2n uses for [client hellos](https://github.com/awslabs/s2n/blob/main/docs/USAGE-GUIDE.md#s2n_config_set_client_hello_cb) as well as [async private key operations](https://github.com/awslabs/s2n/blob/main/docs/USAGE-GUIDE.md#asynchronous-private-key-operations-related-calls). However my feeling is that the ergonomics with callbacks would be poor with Rust, due to lifetime concerns. Additional accessors for retrieving handshake information and requirements should work better.

**Comment by Matthias247** [comment]
Instead of having an explicit `set_cert` method the existing [`ResolvesServerCert`](https://docs.rs/rustls/0.19.0/rustls/trait.ResolvesServerCert.html) could also continued to be used. Users would need to implement it in a fashion which initially returns "cert not available", which would pause the handshake and e.g. set the `wants_cert` flag described earlier. Users can then read the SNI, update their resolvers state to make the cert available, and retry. I guess this would require to be able to return something like [`Poll::Pending`](https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Pending) from `resolve()`, which isn't supported by the current API.

**Comment by Ralith** [comment]
I think having the state transitions explicit in the types makes this a bit less error prone, since there's no possibility of confusion between a connection that's still reading the `ClientHello` and one that's been fully configured. Still, it'd definitely be an improvement over the synchronous status quo.

**Comment by Rantanen** [comment]
While the issue title relates to certificate resolution, there are other use cases brought up in the discussion for intercepting the `ClientHello`, such as figuring out what kind of extension data should be sent in the `ServerHello`. A solution that allows access to `ClientHello` early enough without assuming too much about what the user wants to do with it would probably serve different use cases better.

**Comment by djc** [comment]
> While the issue title relates to certificate resolution, there are other use cases brought up in the discussion for intercepting the `ClientHello`, such as figuring out what kind of extension data should be sent in the ServerHello. A solution that allows access to `ClientHello` early enough without assuming too much about what the user wants to do with it would probably serve different use cases better.

What data from the `ClientHello` would this depend on?

**Comment by Matthias247** [comment]
One example which currently also doesn’t seem to be supported is configuring the ALPN response based on the ClientHello. Multi-tenant servers (like CDNs) can’t use a static ALPN, since eG individual hosted sub-services can have different preferences - like enabling or disabling HTTP/2 support. 

Those services can also require setting The list of supported TLS versions and cipher suites based on the SNI.

**Comment by djc** [comment]
I actually have an initial attempt at starting to fix this, along the lines of the API proposal in https://github.com/ctz/rustls/issues/89#issuecomment-631042052. I was trying to share as much code as possible with the "normal" path, only pulling in extraction of the data required for the `ClientHello` interface. Hope to continue on this soon (but don't block on my account).

**Comment by Rantanen** [comment]
> What data from the `ClientHello` would this depend on?

[My use case](https://github.com/Rantanen/proxide/blob/829d4b37af81c89485cfae50295e3d8450848d0f/src/connection/tls.rs#L43-L54) required ALPN in addition to SNI. Essentially I need `ClientHello` from the client, so I can send a similar `ClientHello` to another server and then read the `ServerHello` back to figure out what kind of a `ServerHello` I need to send back to the original client.

In a perfect world a transparent proxy would need "all of ClientHello", but I'm currently paying attention to only ALPN and SNI since those are the two things my use case of intercepting gRPC required.

**Comment by jsha** [comment]
I'm interested in this functionality (choosing both ALPN and certificate based on the contents of ClientHello) for the crustls C-to-rustls bindings (https://github.com/abetterinternet/crustls), which will be used in Apache and hopefully other web servers. To add the the reasons above for avoiding Futures: exporting the concept of Futures to C would add complexity.

I also wanted to link this issue to the cool work @djc is doing in https://github.com/ctz/rustls/pull/443. Maybe there's some useful synergy here? That is, construction of `Acceptor` doesn't need randomness. It's when we create a `ServerSession` from an `Acceptor` (or more specifically when we first do `write_tls` from that `ServerSession`) that we need to get random bytes and possibly fail.

I like the interface suggested in https://github.com/ctz/rustls/issues/89#issuecomment-630776232. My original inclination was a `wants_cert` like suggested in https://github.com/ctz/rustls/issues/89#issuecomment-751957552, but the `Acceptor` model seems like a nice [application of the Typestate pattern](http://cliffle.com/blog/rust-typestate/) and also potentially allows using less memory for connections that haven't finished their handshake.

**Comment by Rantanen** [comment]
This would also be needed to support certificate/cipher selection based on client capabilities. The server might want to allow EC-certificates for modern clients, but still support RSA-certificates for older clients that do not have support for EC ciphers.

This would require the server to examine the client ciphers (As far as I remember, Client advertises supported ciphers in ClientHello?) and then select a compatible certificate based on those.

**Comment by ctz** [comment]
> This would also be needed to support certificate/cipher selection based on client capabilities. The server might want to allow EC-certificates for modern clients, but still support RSA-certificates for older clients that do not have support for EC ciphers.

This is already possible with the existing ResolvesServerCert trait -- see https://docs.rs/rustls/0.16.0/rustls/trait.ResolvesServerCert.html#tymethod.resolve

**Comment by Rantanen** [comment]
Oh, in that case disregard whatever I said about ciphers above. :)

**Comment by pickfire** [comment]
> My use case required ALPN in addition to SNI. Essentially I need ClientHello from the client, so I can send a similar ClientHello to another server and then read the ServerHello back to figure out what kind of a ServerHello I need to send back to the original client.

I have the same issue and as it isn't async, I need to change the code structure again. T_T

**Comment by djc** [comment]
I have an implementation of this idea in #787.

**Comment by AlbertMarashi** [comment]
any ETA?

**Comment by djc** [comment]
@AlbertMarashi this is available in 0.20, has been for a few months now.

**Comment by AlbertMarashi** [comment]
@djc anywhere I could find an example or documentation?

**Comment by djc** [comment]
There is the reference documentation. There's also a `LazyConfigAcceptor` in tokio-rustls. If you have specific questions or concerns, please open a new issue.

**Comment by jsha** [comment]
@AlbertMarashi I've expanded the reference documentation in #1241, along with an example. It will be in an upcoming release.
